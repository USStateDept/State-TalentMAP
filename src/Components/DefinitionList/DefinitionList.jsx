import React from 'react';
import PropTypes from 'prop-types';
import shortid from 'shortid';
import { isArray, isPlainObject, keys, map, merge, omit } from 'lodash';
import Definition from './Definition/Definition';

const defaults = {
  items: [],
  children: [],
  truncate: true,
};

const DefinitionList = ({ truncate = true, ...props }) => {
  const options = omit(props, keys(defaults));
  const children = !isArray(props.children) ? [props.children] : props.children;
  let items = props.items;

  // Transform prop type objects
  if (isPlainObject(items)) {
    items = map(items, (item, key) => merge({
      term: key,
      definition: item,
      truncate,
    }));
  }

  options.className = options.className ? `definitions ${options.className}` : 'definitions';

  return (
    <dl {...options}>
      {
        children.length ?
          children :
          items.map(item => (
            <Definition key={shortid.generate()} {...item} />
          ))
       }
    </dl>
  );
};

const node = PropTypes.shape({
  type: PropTypes.oneOf([Definition]),
});

DefinitionList.propTypes = {
  /** Ignored if `children` if both props are used */
  items: PropTypes.oneOfType([
    /** Object: keys => terms, values => definitions */
    PropTypes.shape(),
    /** Array of { term: term, definition: definition } */
    PropTypes.arrayOf(
      PropTypes.shape({
        term: PropTypes.string,
        definition: PropTypes.string,
        truncate: PropTypes.bool,
      }),
    ),
  ]),

  /** Takes precedence over `items` if both props are used. */
  children: PropTypes.oneOfType([
    PropTypes.arrayOf(node),
    node,
  ]),

  truncate: PropTypes.bool,
};

DefinitionList.defaultProps = defaults;

export default DefinitionList;
