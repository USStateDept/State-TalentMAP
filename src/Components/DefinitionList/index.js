import DefinitionList from './DefinitionList';

export { default as Definition } from './Definition/Definition';
export default DefinitionList;
