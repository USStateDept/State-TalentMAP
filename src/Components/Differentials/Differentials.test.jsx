import { shallow } from 'enzyme';
import toJSON from 'enzyme-to-json';
import DefinitionList from './Differentials';

describe('DefinitionList', () => {
  it('matches snapshot', () => {
    const wrapper = shallow(<DefinitionList />);
    expect(toJSON(wrapper)).toMatchSnapshot();
  });
});
