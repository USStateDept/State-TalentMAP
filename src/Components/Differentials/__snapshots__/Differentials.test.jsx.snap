// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DefinitionList matches snapshot 1`] = `
<div>
  <span>
    N/A
     | 
  </span>
  <span>
    N/A
  </span>
  <span
    style={
      Object {
        "paddingLeft": 5,
        "position": "absolute",
      }
    }
  >
    <OBCUrl
      label="OBC Details"
      type="post-data"
      url={Object {}}
    />
  </span>
</div>
`;
